### ISSUE TITLE
[FR] Improvements regarding OIDC

### ISSUE THREAD

**Issue by Westie** [docker-mailserver/docker-mailserver#2713]
I've tagged this as an issue as I've had to completely hack around with various parts of this docker project to get what I want it to do with regards to loading stuff how I want it to load, and I thought it would probably be nice to get this made far nicer.

I'd like to volunteer myself to at least partly solve this issue as I want to make what I've done far easier to main - at present I'm relocating init scripts in an attempt to modify fewer files.

c.f.: <https://github.com/docker-mailserver/docker-mailserver/issues/2312>

## Life so far

```
COPY conf/dovecot/09-auth.conf /etc/dovecot/conf.d/09-auth.conf
COPY conf/dovecot/auth-ldap.conf.ext /etc/dovecot/conf.d/auth-ldap.conf.ext
COPY conf/dovecot/auth-oauth2.conf.ext /etc/dovecot/conf.d/auth-oauth2.conf.ext
COPY conf/dovecot/dovecot-oauth2.conf.ext /etc/dovecot/dovecot-oauth2.conf.ext
```

### `scripts/setup-stack.sh`

Within my hijacked `start-mailserver.sh` I have the following:

```
configomat.sh OAUTH_ /etc/dovecot/dovecot-oauth2.conf.ext
```

Notwithstanding that I realise that this should actually be `OAUTH2_` (sue me, I wasn't thinking) - this is a rather easy addition to the project as we can just stick it within [setup-stack.sh](https://github.com/docker-mailserver/docker-mailserver/blob/master/target/scripts/startup/setup-stack.sh) somewhere.

### `dovecot/10-auth.conf`

At present I'm doing weird magic in order to ensure that the OAUTH2 driver is being loaded before the LDAP driver i.e., I'm creating a separate file called `09-auth.conf` and inserting my required config entries in there.

I had tried setting LDAP to specific authentication mechanisms but internally rewritten e-mails stopped getting delivered.

The project has its own `10-auth.conf` [file](https://github.com/docker-mailserver/docker-mailserver/blob/master/target/dovecot/10-auth.conf) which has the following order:

```
#!include auth-system.conf.ext
#!include auth-sql.conf.ext
#!include auth-ldap.conf.ext
!include auth-passwdfile.inc
#!include auth-checkpassword.conf.ext
#!include auth-vpopmail.conf.ext
#!include auth-static.conf.ext
```

Where would we put the reference for `auth-oauth2.conf.ext`? I'd personally rank it above LDAP but below SQL.

### `dovecot/auth-oauth2.conf.ext`

```
passdb {
    driver = oauth2
    mechanisms = oauthbearer xoauth2
    args = /etc/dovecot/dovecot-oauth2.conf.ext
    skip = authenticated
}
```

### `dovecot/dovecot-oauth2.conf.ext`

```
client_id = ?
client_secret = ?
grant_url = https://localhost/grant
introspection_mode = post
introspection_url = https://localhost/introspection
tls_ca_cert_file = /etc/ssl/certs/ca-certificates.crt
```

### Other improvements

In my copy of `dovecot-ldap.conf.ext` I have removed the LDAP userdb and I've decided to solely rely on a static userdb entry instead.

That's probably a silly mistake on my behalf but what is a good idea to add to that file is something like `skip = authenticated` or even something `skip =` so that this can be modified out of the box via an environment variable

## So, what do?

I think this is what we'll need to do:

* Implement an `ENABLE_OAUTH2` that will uncomment and include the OAUTH2 driver
* Ensure that OAUTH2 is loaded before LDAP to avoid un-necessary failed auth attempts (it will shout at you because the JWT is unsurprisingly too long for LDAP to comprehend)

Seem fair?

**Comment by georglauterbach** [comment]
I am willing to help you with this (if some conditions are met, see below), I have the necessary script knowledge... I would like to have some questions answered beforehand :)

1. Does LDAP have to be enabled for OIDC (i.e. OAuth2..) to work? If this is the case, I think I am of little help as I do _not_ have any experience with LDAP at all really. This is probably the most important question.
2. Can you elaborate more on how aliases work with OIDC? How would "multiple email addresses for one account" work?
3. Do we need configomat (in the scripts)? I would like to get rid of configomat in the future, and using it here would make life unnecessarily harder.

I have Authelia running in my own infrastructure, which provides OIDC services.

**Comment by polarathene** [comment]
I can also offer advice for a large scope of the DMS project and getting the feature developed. I would like to see OIDC support as well.

> 1. Does LDAP have to be enabled for OIDC (i.e. OAuth2..) to work? If this is the case, I think I am of little help as I do _not_ have any experience with LDAP at all really. This is probably the most important question.

My understanding was they just wanted it for the PassDB authentication, while providing UserDB queries via LDAP.

Presently our Dovecot config ships both PassDB and UserDB.

For LDAP I recently looked into what's going on there and realized that we swap Dovecot SASL for SaslAuthD service instead (with Auth Bind enabled), but Dovecot provides LDAP auth for it's protocols IMAP/POP3 (without Auth Bind enabled) instead of the SaslAuthD service (Postfix only). That was due to the original LDAP auth support being contributed by someone that wasn't interested in using Dovecot for it. I'd like to look into refactoring that at some point once I get time for looking into LDAP support.

For this feature AFAIK, splitting out UserDB and PassDB configs to their own separate files and toggling those would work just as well.

I'm working on resuming the encryption PR at present, but ran into an issue with this Dovecot config management myself, where I'll follow-up with another PR to co-locate logic that swaps these configs between our default and LDAP, so any OIDC changes could probably be bundled there too.

> Do we need configomat (in the scripts)? I would like to get rid of configomat in the future, and using it here would make life unnecessarily harder.

I assume it's for convenience. We definitely don't want to have more explicit ENV. Personally I'd rather a sane default if possible and deferring to config mounting/replacement like we do elsewhere instead of micro-managing modifications with ENV and sed.

Likewise, another `ENABLE_` ENV for a different auth backend would be nice to avoid, but until we have a better approach to consolidate that handling I can approve it. `ENABLE_LDAP` is more intrusive through our scripts, whereas the OIDC feature discussed here doesn't seem to be fairly minimal, warranting the need for the ENV even less.

---

If we do deprecate ENV for such, we could still support it via `setup` or a `user-patches.sh` example folder for docs to reference. As we're now talking about supporting multiple UserDB/PassDB it would make more sense to rely more on files than wading through ENV I think (_which has caused a few maintenance gripes via configomat_).

From what I've seen, the LDAP feature requires a bunch of ENV lines, which for the most part defers to config files already and once those are modified to match the users config needs, it's fairly static. They could produce the config files the same way without muddying DMS ENV and scripts (it'd be nice to drop the submodule). So I'm in favour of stripping that out as another preparation step for setting up DMS, I've already got the rust port of `configomat.sh` I could publish, or put together something similar for our specific config needs instead of associated shell scripts that adapt those configs and ENV to the LDAP config requirements/generation.

---

> We can surely test a future implementation with other services as well.

This is something I would expect for approval. A test must be contributed with OIDC setup, not just local confirmation of the feature working.

Once the feature is merged, maintenance burden is unloaded onto us, and we already get plenty of LDAP issues, but at least our tests help somewhat ensure compatibility. I don't really want to see OIDC added without tests, but I understand if it'd be rather difficult to test, if it is merged without tests it is less likely to be kept around if it causes maintenance burden going forward.

**Comment by georglauterbach** [comment]
> > Do we need configomat (in the scripts)? I would like to get rid of configomat in the future, and using it here would make life unnecessarily harder.
>
> I assume it's for convenience. We definitely don't want to have more explicit ENV. Personally I'd rather a sane default if possible and deferring to config mounting/replacement like we do elsewhere instead of micro-managing modifications with ENV and sed.
> 
> 
> 
> Likewise, another `ENABLE_` ENV for a different auth backend would be nice to avoid, but until we have a better approach to consolidate that handling I can approve it. `ENABLE_LDAP` is more intrusive through our scripts, whereas the OIDC feature discussed here doesn't seem to be fairly minimal, warranting the need for the ENV even less.

I am actually currently working on a PR which reworks the user provision mechanism variables. This is a bigger change, but it should be backwards compatible the way I design it. With `v12.0.0.` we can then remove the compatibility code.

**Comment by Westie** [comment]
> Do we need configomat (in the scripts)? I would like to get rid of configomat in the future, and using it here would make life unnecessarily harder.

I heavily rely on setting environment variables to configure my mailserver so I'd love to see what you have as a replacement

> Once the feature is merged, maintenance burden is unloaded onto us, and we already get plenty of LDAP issues

This may be a problem, setting up an identity provider can be a bit of a pain. I am not too sure if Keycloak can be used to automate the process but I suspect something considerably lighter weight such as Authentik could be used possibly.

**Comment by Westie** [comment]
> Does LDAP have to be enabled for OIDC (i.e. OAuth2..) to work? If this is the case, I think I am of little help as I do not have any experience with LDAP at all really. This is probably the most important question.

You don't need to have LDAP to have OIDC, but the pair of them practically go hand in hand. One scenario, if it was available to you, could be to connect directly to a SQL server powering the identity provider and directly query whether or not the user exists.

**Comment by Westie** [comment]
> For LDAP I recently looked into what's going on there and realized that we swap Dovecot SASL for SaslAuthD service instead (with Auth Bind enabled), but Dovecot provides LDAP auth for it's protocols IMAP/POP3 (without Auth Bind enabled) instead of the SaslAuthD service (Postfix only). That was due to the original LDAP auth support being contributed by someone that wasn't interested in using Dovecot for it. I'd like to look into refactoring that at some point once I get time for looking into LDAP support.

Could you clarify that please?

In my setup, Postfix authentication is delegated to Dovecot and Dovecot uses bound authentication

**Comment by polarathene** [comment]
> Could you clarify that please?

Sure, there is an open issue about an Active Directory LDAP user having trouble, where I comment with a [deep dive on LDAP config / logic here](https://github.com/docker-mailserver/docker-mailserver/issues/2711#issuecomment-1206498746).

Active Directory appears to only be compatible with `auth_bind = yes` which our Dovecot LDAP config does not enable, and we appear to lack documentation on how to switch to that. This would only affect auth from mail clients for IMAP4/POP3 protocols to retrieve mail. Sending mail over ports 25/587/465 is Postfix with the last two of those ports being SASL auth related.

In a non-LDAP config, our Postfix defers SASL auth provider to Dovecot, but the original contributor of our LDAP support did not want to use Dovecot and instead introduced Cyrus SASL implementation `saslauthd` for Postfix to be configured with instead.

This allows for `saslauthd` to be configured for LDAP (_which is the only config we explicitly configure fallback config for with `LDAP_*` and `DOVECOT_*` ENV using `configomat`_), that is by default configured for Auth Bind, while ports/protocols that Dovecot manages uses it's own internal LDAP support for `PassDB` and `UserDB` for it to perform queries, however the PassDB not using Auth Bind expects an LDAP query to return `userPassword` attribute AFAIK.

`saslauthd` was chosen by the contributor as they had no interest in Dovecot (that LDAP support came later I think), they wanted to pair DMS with Kopano. The `saslauthd` feature supports other configs beyond LDAP such as `rimap` which sends login credentials to an IMAP service to authenticate with, and we appear to have a test for that, but I didn't look further into that.

---

You'll find our LDAP support has not been aging well due to lack of LDAP experienced maintainers.

- The tests still use a very dated pinned version of an OpenLDAP container IIRC.
- Having mail accounts use separate UID/GID ownership assigned to `/var/mail` user folders is not supported at present.
- SpamAssassin / Amavis has an issue (should be open) specific to LDAP users, the fix shouldn't be too difficult IIRC, I just lack time.
- The mixed SASL providers mentioned above.
- The default LDAP config not supporting Auth Bind in Dovecot for Active Directory LDAP users.
- Possibly other concerns.

**Comment by polarathene** [comment]
> You don't need to have LDAP to have OIDC, but the pair of them practically go hand in hand.

I assume we can leverage a `passwd-file` UserDB like we do by default, and just delegate PassDB driver to OIDC for auth?

I understand that probably wouldn't make much sense though, due to likely duplicating managements of user accounts that are likely via an IAM? Hence the decision to pair with LDAP?

**Comment by polarathene** [comment]
> I heavily rely on setting environment variables to configure my mailserver (I have it set up so that I have prod and staging environments) so I'd love to see what you have as a replacement

I believe the ENV with `configomat` being discussed is the main concern. It's only being used to modify config files, with some additional logic on our part to derive defaults (when some ENV are not specified and/or referencing other ENV).

Nothing would stop you from using that externally of DMS, but effectively you'd just be mounting two different config files. ENV can be used in `docker` CLI and `docker-compose.yml` for changing the value of paths you assign to a volume mount.

**Comment by Westie** [comment]
> Sure, there is an open issue about an Active Directory LDAP user having trouble, where I comment with a https://github.com/docker-mailserver/docker-mailserver/issues/2711#issuecomment-1206498746.

Wow, LDAP in this project is a mess 😂 

I thought I had to do special stuff to get LDAP working because of special ways that I was doing it, however I think it's more that this project needs a review of LDAP config, possibly

**Comment by georglauterbach** [comment]
We would **love** to have someone to look at our LDAP code :D Core maintainers do not use LDAP and can therefore not really maintain it - it is therefore (with certainty) an absolute mess.

Maybe you could help us in this regard @Westie?

**Comment by Westie** [comment]
I'd like to help out but I can't commit to a definitive timeframe unfortunately.

I suspect it may just be a case of one night I download a fresh copy of the repo and go to town on it, really.

**Comment by georglauterbach** [comment]
> I'd like to help out but I can't commit to a definitive timeframe to review both LDAP and OIDC unfortunately.
> 
> I suspect it may just be a case of one night I download a fresh copy of the repo and go to town on it, really. 

That would be awesome 👍🏼

**Comment by polarathene** [comment]
No rush. Probably best to wait until after v12 release if it's going to be invasive changes.

Ping either of us if you find some time and need any information / guidance about the project :+1:

**Comment by Westie** [comment]
> Probably best to wait until after v12 release

I agree.

**Comment by Westie** [comment]
@polarathene / @georglauterbach I see that there's been a minor release, is there a timeline for v12? life is getting busy right now and i'd like to figure out what personal time i can allocate and when

**Comment by Westie** [comment]
https://github.com/docker-mailserver/docker-mailserver/blob/ab55343d8ee34ccdf562804b2c098a9938b026c8/target/scripts/start-mailserver.sh#L73

@georglauterbach - this is interesting behaviour, as in the real world you will need both.

LDAP for clients such as Apple Mail and OIDC for clients such as Roundcube et al. How do you intend to allow both?

**Comment by Westie** [comment]
In addition the setup that I described in the OP has been largely unchanged and has proven itself to be very stable.

**Comment by georglauterbach** [comment]
Wouldn't it be possible to either enable LDAP implicitly when enabling OIDC, or have a new toggle `LDAP+OIDC`?

---

v12 will be the next release in case a breaking is merged before v11.3.0 is released; there is no fixed schedule.

**Comment by polarathene** [comment]
@Westie I believe there's a `v11.3.0` scheduled, but no major changes there, feel free to branch off the tagged release commit on master branch and get a PR going.

I've got limited time available until December, so anything I have planned to do is on hold and will be rebased for a future release. When I can spare the time I am checking in on issues or PRs to help with reviews.

We should be able to review and merge your contribution. Once a PR (_it could be more than one if helpful_) is up, we'll see how to approach it. If it's a large one we might try split it up into smaller PRs to get those reviewed and merged more easily.

---

> @georglauterbach - this is interesting behaviour, as in the real world you will need both.

I suppose it depends on what your requirements are for deployment. Some deployments may be comfortable with just offering webmail with Roundcube as the userbase may be small / specific enough that it's suitable? Similar to LDAP deployments that don't use OIDC now?

If that is something you'd like to improve, it would be a good case for a separate PR to focus on, but if the feature PR is small enough, I'm open to it all being in one PR.

Are you going to have enough time to contribute towards docs or tests? We can try assist with tackling that once the core feature itself is opened as a PR. Both would be great to have, but understandably raise the contribution effort required :sweat_smile:

**Comment by Westie** [comment]
> If that is something you'd like to improve, it would be a good case for a separate PR to focus on, but if the feature PR is small enough, I'm open to it all being in one PR.

it could possibly be `ACCOUNT_PROVISIONER=LDAP,OIDC` where the string is split into an array and the resulting array looped through

**Comment by Westie** [comment]
> Are you going to have enough time to contribute towards docs or tests? We can try assist with tackling that once the core feature itself is opened as a PR. Both would be great to have, but understandably raise the contribution effort required

Given the little support on offer for LDAP, I think that yes, ultimately I would need to include those as part of that process. Not necessarily to find bugs but I think more importantly to provide a way to share what madness was going on in my brain  whenever I decided to do what I would do.

**Comment by fbartels** [comment]
> and OIDC for clients such as Roundcube et al

Maybe a bit off topic, but have you encountered any mail clients apart from Roundcube that could actually make use of this functionality? Thunderbird does have support for oauth for IMAP/SMTP, but they opted to hard code the identity providers so its only good when using e.g. Google as the identity provider.

**Comment by Westie** [comment]
@fbartels I wouldn't consider it off-topic at all.

Desktop or mobile clients - as far as I'm aware - generally do not support authentication over XOAUTH2/OAUTHBEARER to custom hosts, however it would be somewhat of a moot point as setting up a desktop mail client is sort of fire and forget - you need to do it only once, hopefully.

**Comment by fbartels** [comment]
@Westie while its true that imap client setup is pretty much "fire and forget" using oauth also comes with some security benefits, like coupling the login at the openid provider with 2fa and of course the fact that the client would not store the actual password, but only a token with a certain lifespan.

**Comment by MohammedNoureldin** [comment]
I am having eyes occasionally on the LDAP code (at least if it works, and reviews some parts when I have time).

I would not focus a lot on LDAP when discussing OIDC, because actually one of the main benefits of using IdP is to keep the multiple authentication methods (or yet better the mixture of methods) abstract. LDAP can be just one of multiple authentication backends provided by IdP. A very common enterprise level example is a workflow that uses LDAP and MFA to authenticate. In the end, Docker Mailserver will know nothing about the backend used by the IdP, but only about the tokens it receives.

**Comment by georglauterbach** [comment]
I see; I am not an expert when it comes to LDAP / OIDC. My main concern in the end is that the code can be properly maintained; I am unable to write the code myself though, so there is no problem adjusting the current state concerning OIDC I think.

**Comment by MohammedNoureldin** [comment]
I can tell about LDAP, but I am also not an expert in OIDC, can anybody start implementing it? I may be able to support when I see that start of the road.

**Comment by Westie** [comment]
> I would not focus a lot on LDAP when discussing OIDC, because actually one of the main benefits of using IdP is to keep [things separate]

Yes, however by excluding LDAP, you end up in the situation where you're not able to authenticate using a desktop mail client as desktop mail client support for OIDC is poor.

I intend to start looking at this over the next month or so @MohammedNoureldin

**Comment by georglauterbach** [comment]
Just a suggestion: with the upcoming v12.0.0, there are a few important and breaking changes. Especially changes to the test suite (actively worked on at the moment) and DNS name handling (todo). v12.0.0 has no clear ETA yet, so if you start working, please bear in mind (and feel free to ask!) to use the most up-to-date version of this repo and rebase periodically.

Feel free to ask me anything about the scripts; other maintainers can help you with other configuration as well :)

**Comment by DiscowZombie** [comment]
Hello,

Is there any update on this? Has some of you already started something usable? I will be highly interested in this feature.

**Comment by polarathene** [comment]
> Has some of you already started something usable?

I wouldn't expect anyone that was interested working on it has started.

There was heavy amounts of restructuring our code for v12. After that is released later this month it should be approachable, but none of the maintainers have time to work on such AFAIK beyond providing reviews / advice.

**Comment by jastlw** [comment]
Hi, 
if this features is released, can I also use Microsoft SMTP with modern authentication (Auth0)?
https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth
Or would this a complete different standard to OIDC?

Thank you in advanced!
Jacob

**Comment by mpldr** [comment]
> can I also use Microsoft SMTP

From experience as a MUA-developer: probably not, as Microsoft decided against Implementing OAuth2, and opted for XOAuth.

**Comment by MohammedNoureldin** [comment]
Hi @Westie, how is it going? I am just poking you, because you said you might be able to start working on this, and I volunteered to support happily :) Have you made any progress? If you not, but still you can begin soon, I am still happy to support. Please let me know.

During that time, I enriched my knowledge a bit more about OIDC to be more helpful ;)

**Comment by MohammedNoureldin** [comment]
Hi, @georglauterbach and @polarathene!

In the past I put my focus to get the LDAP working, but actually what I found is that the focus should be on SSO (SAML 2.0 or OIDC) not LDAP (where LDAP is still nice to have). SSO is a standard for enterprise, and having such feature would be great and kind of deal breaker, I found that mailcow started working on it now and they would release it somewhen this year, that is to say that it should also be possible somehow for this project. I am still not sure how the different clients on different platforms will deal with SSO. Still, this is something to be analyze.

I am also very beginner regarding implementing SSO in code, but I think if multiple people work together on this, we will be able to get it working somewhen, though the pain :D.

What do we need to implement it? I would be glad to help as far as possible.

@Westie sorry for pinging you again, I wanted to be sure that we clarify the current situation, maybe you started something in this regard?

**Comment by georglauterbach** [comment]
> I am also very beginner regarding implementing SSO in code, but I think if multiple people work together on this, we will be able to get it working somewhen, though the pain :D.

A good idea, but all maintainers (including me) are so low on time that starting something new for them is just infeasible at the moment :(

> What do we need to implement it? I would be glad to help as far as possible.

I have no clue about SSO. And you will probably need to start working yourself, because this seems to have stalled a bit (I guess?).

---

My recommendation: providing small, incremental PRs and having a proper tracking issue should do the trick. This is easy and (somewhat) fast for maintainers to review and then merge.

**Comment by polarathene** [comment]
> What do we need to implement it? I would be glad to help as far as possible.

- You can run the container and shell into it with `docker exec -it container_name bash`.
- You may want a text editor in the CLI to edit files inside the container, you can use `apt-get update && apt-get install nano`. Then you just use the `nano` command and a file path to create / edit a file.
- Use a search engine or read through docs of Dovecot, learn how to configure OIDC. You also have some context in this issue, especially from the author detailing some related config.
- You can restart dovecot with `supervisorctl restart dovecot`, or probably better is to try reloading with `dovecot reload`. That should apply the config changes / additions.

Once you get that working and connect successfully with OIDC, then you'd look at how to integrate that into DMS via scripts and such, which maintainers can respond to any questions such as locations to add such. Feel free to put a PR forward and discuss it then, doesn't have to be perfect :)

I haven't reviewed the earlier discussion here, so the whole config / setup part may have already been resolved and just needs integration scripts / ENV + docs for DMS.

**Comment by MohammedNoureldin** [comment]
@polarathene is it only about Dovecot? What about the other parts, particularly Postfix?

**Comment by polarathene** [comment]
> What about the other parts, particularly Postfix? Could you explain a bit more about this?

- Mail storage is handled via Dovecot over IMAP and POP3 protocols, that requires authentication.
- Mail submission via SMTP on ports 587 and 465 is handled by Postfix, but it delegates authentication to SASL provider.
- DMS is configured for SASL accounts via Dovecot, although we provide an alternative provider with SASLAuthd if a user enables that instead. IIRC we have tests only for `ldap` and `rimap` providers of SASLAuthd.
- This feature request discusses OIDC support from Dovecot (_or rather OAuth2 it seems at a quick glance_).

**Comment by TBG-FR** [comment]
Here's my take on the subject. I'm not really a mail expert, but given the short amount of time I had to find a working solution, I thought it would be interesting to share my thoughts and solutions, in order to help others (and maybe contribute to the `OIDC` provider development).

## Enable OIDC (OAuth2) authentication in Docker Mailserver
Following your comments here and some guides such as https://doc.dovecot.org/configuration_manual/authentication/oauth2/ you can easily configure Dovecot to accept authentication from your OIDC/OAuth2 provider (in my case, Keycloak)

<details>

`/etc/dovecot/conf.d/10-auth.conf`
```diff
-auth_mechanisms = plain login
+auth_mechanisms = xoauth2 oauthbearer plain login

+!include auth-oauth2.inc
```

`/etc/dovecot/conf.d/auth-oauth2.inc`
```
passdb {
  driver = oauth2
  mechanisms = xoauth2 oauthbearer

  # Path for OAuth2 configuration file
  args = /etc/dovecot/dovecot-oauth2.conf.ext
}
```

`/etc/dovecot/dovecot-oauth2.conf.ext`
```
introspection_url = ...
introspection_mode = post
client_id = ...
client_secret = ...
...
```
</details>

However, at this point **account login will only work if you previously created this account** (I'm using `ACCOUNT_PROVISIONER=FILE` btw). So we need something that will be able to create accounts (= fill `postfix-accounts.cf` and wait for the `changedetector` to trigger `accounts.sh` & all)

## Effectively create accounts

### Alternative 1 - Using a static userdb

Using the `userdb` config for `passwd-file` and the configuration provided at https://documentation.open-xchange.com/7.10.4/middleware/mail/dovecot/oauth_2.0_with_postfix_and_dovecot.html we added a `userdb` entry to the `oauth2` config file. Doing so, we were able to connect successfully (using Roundcube for example), and we could send emails, but that new account couldn't be found by others (some files were missing because not created, obviously).

`/etc/dovecot/conf.d/auth-oauth2.inc`
```diff
+userdb {
+  driver = static
+  args = username_format=%u uid=docker gid=docker home=/var/mail/%d/%u
+  default_fields = uid=docker gid=docker home=/var/mail/%d/%u
+}
```

**This first alternative is not fully satisfying because it does not setup everything correctly.**

### Alternative 2 - Using post-login scripting

I believe we could use dovecot's [post-login scripting](https://doc.dovecot.org/admin_manual/post_login_scripting/) to trigger a modification inside the `postfix-accounts.cf` with a simple shell script and wait for the folder/files to be created.

**This second alternative feels satisfying.** An implementation to validate it would be interesting. However, that raise a new issue : **mail accounts/boxes will not exist prior to the first login of each user** 🤯 So you will not be able to receive any email, even if your account exists in your OIDC Provider, because on the mailserver side, nothing exists.

### Alternative 3 - External service

In order to adress the two issues (accounts need to be created to be able to login, accounts need to be created to receive emails on existing accounts before login), another alternative is to use an external service/trigger/hook, that would (regularly or on each new user creation) get existing users from the provider and create them in mailserver using `setup email add`.

**That's not perfect because it's an hybrid mode (between `FILE` and the goal for `OIDC` I believe), and it requires an external custom component, but that can solve (specific, I know) issues.**

## The end

Do not hesitate to give me your thoughts on these elements, and correct me if I'm wrong 😉

**Comment by polarathene** [comment]
> Do you know what about LDAP implementation now in Docker mailserver, does it do the part you mentioned?

If you use an identity broker like Keycloak, they can integrate LDAP for account management IIRC, while the IDP like Keycloak can handle OIDC.

I am not too familiar with LDAP, and it may stiill need integration with DMS, but that may no longer be relevant with the OIDC provider :man_shrugging: 

> If not, maybe we can check how LDAP works at the moment?

DMS supports either the file based configs for accounts/config, or LDAP for sourcing the same information IIRC.

I don't have experience with OIDC and integration with Dovecot/Postfix to know if it can be queried as a data source, but from the configs shown above where it only seems to be used for auth, I'm guessing Dovecot and Postfix aren't able to query for such. I am familiar that OIDC can provide some user metadata, but it's probably not the same :sweat_smile: 

Hence the file configs.

---

@TBG-FR Thanks for the awesome write-up, really appreciate it :+1: 

I'm not able to commit much time to discussing this topic any time soon, and I think that's the same for the other maintainers right now unfortunately.

Not every user from an OIDC provider should necessarily get an account on DMS. In a scenario where I would use it, it'd only be relevant to admins, not general users of a community with their own personal emails but both admin and users of the community using Keycloak for SSO (_I never got around to implementing this, so they all login via a social login instead_).

I believe the LDAP support works in a similar way, and that it's flexible at controlling what users can have mail accounts with DMS.

I wouldn't consider the 2nd solution desirable, if you create an account for an employee and they're assigned an email address with DMS, then it would be ideal that they could have mail sent to them prior to their first login. You're welcome to choose that via PR and document the caveat, but it'd be great if it would instead match the experience of `FILE` and `LDAP` provisioners?

The third solution, an API would be nice and has been talked up in the past but no one has had time to properly contribute one. That said, manual account creation with `setup email add` could still be valid until someone does have time to introduce an API. Perhaps it's the best approach for the time being to get OIDC supported?

**Comment by MohammedNoureldin** [comment]
Yes I use Keycloak with LDAP integration. If OIDC implementation supports the whole things like aliases, groups, etc., this would be awesome!

Actually it makes much more sense to keep sourcing the information without storing any new account in DMS. Exactly like LDAP at the moment. This keeps DMS only a store for mails, not accounts (which keeps everything clean and simple).

**Comment by polarathene** [comment]
> Actually it makes much more sense to keep sourcing the information without storing any new account in DMS. Exactly like LDAP at the moment.

So then we have two different responsibilities that `FILE` and `LDAP` handle, and `OIDC` would replace one of those for both `FILE` and `LDAP`, rather than be it's own full replacement?

So minor refactoring perhaps would better tailor to that distinction? Then account creation is a separate concern which the earlier API discussion/attempt would support.

**Comment by NhProGamer** [comment]
is OIDC login usable with a LDAP backend ?

**Comment by polarathene** [comment]
@NhProGamer yes, although you might have to wait until v15 when LDAP support is reworked, the OAuth2 / OIDC support will leverage the same improvements to make it easier to configure for.

The way the support of this feature works for auth is `XOAUTH2` / `OAUTHBEARER` auth provides an access token. That is checked against the auth service to get additional information such as an email attribute (default), which is then compared by Dovecot against the login username provided (typically an email address in DMS).

I had tested it against LDAP myself and it worked, so if it doesn't work for you it might just need some minor config changes.

**Comment by Westie** [comment]
@polarathene @georglauterbach firstly, I'd like to apologise for the delay that it's taken me to get back to this issue.

Life has unfortunately got in the way, I suspect my loss of interest was caused somewhat by that delay between getting configomat deprecated and it being replaced with the self-hosted method of substituting environment variables.

But first, I'd like to thank the pair of you for sorting most of this out, it's nice to see this mostly make its way to being something that is production ready.

Anyhow, now that life has given me the opportunity to upgrade from [v11.1.0](https://github.com/docker-mailserver/docker-mailserver/tree/v11.1.0) to [v13.3.1](https://github.com/docker-mailserver/docker-mailserver/tree/v13.3.1), I've come back to report any shortcomings - a journey that, I admit, was not fun due to the level of customisation I had in place to support XOAUTH2, but, I've made it!

## Breaking change to dovecot authentication
It turns out that the devs over at the dovecot project made a [really quiet change](https://www.mail-archive.com/dovecot@dovecot.org/msg88463.html) to how introspection authentication occurs. I previously used dovecot's previously supported functionality of POST'ing credentials however the client secret is now required to be in the format of HTTP basic authentication.

## Ordering of auth mechanisms
In my original post, I had posted this:

![](https://ss.westie.sh/VjAVf)

Ultimately that appears because I was seeing errors in the mailserver log such as this:
```
"ldap_bind() failed: Constraint violation"
```

Which of course, makes sense, as what's happening is dovecot is attempting to throw OAUTH2 tokens at a LDAP service and that's _never_ going to end well.

My build doesn't have auth_mechanisms set to include oauthbearer or xoauth2, however I've attempted to include it and switched around the order of auth_mechanisms, so it's something like this:

```
auth_mechanisms = oauthbearer xoauth2 $auth_mechanisms
```

...which doesn't stop those LDAP console messages.

The only reliable way I've found was to ensure that the xoauth2 module was loaded in before LDAP. In the case of this project, the list at the bottom of `10-auth.conf` needs re-ordering but I just stuck it into a file that's loaded first. If you've found any better ways of ordering auth mechanisms, please let me know.

This unfortunately means that I have to have `ENABLE_OAUTH2` set to a falsy value and I have to continue to hijack the init scripts in order to inject environment variables into my custom OAUTH ext file.

## Other config files
### `auth-ldap.conf.ext`
My version of [auth-ldap.conf.ext](https://github.com/docker-mailserver/docker-mailserver/blob/master/target/dovecot/auth-ldap.conf.ext) is simpler than what's on this repo, here is my file:

```
passdb {
    driver = ldap
    args = /etc/dovecot/dovecot-ldap.conf.ext
    skip = authenticated
}

userdb {
    driver = ldap
    args = /etc/dovecot/dovecot-ldap.conf.ext
    skip = found
}
```

Note the presence of `skip`, however I'm unsure whether or not any additional args that are in your copy is required.

### `dovecot-oauth2.conf.ext`
My version of [dovecot-oauth2.conf.ext](https://github.com/docker-mailserver/docker-mailserver/blob/master/target/dovecot/dovecot-oauth2.conf.ext) is almost similar to what's on this repo, here is my file:
```
grant_url = https://localhost
introspection_mode = post
introspection_url = https://localhost

# note: custom TLS options removed
```

I think this file is pretty much custom anyway, I suspect it's going to be rather painful to have all situations catered for within this file.

## Summary
There's actually not much that needs changing to make this into a fully featured, LDAP/XOAUTH2 compatible mail suite. I'll look at creating a PR over the next couple of days to see if that's something you folks may be interested in.

**Comment by polarathene** [comment]
> I'll look at creating a PR over the next couple of days

Before you do, what exactly do you want to PR?

---

I am juggling quite a bit and the LDAP PR that's been open for a while is mostly blocked by my time to rewrite docs (_which I'm deathly slow at_), whilst I juggle other commitments in or out of DMS. However once that PR is merged, the same logic can be used for this oauth config rather easily/quickly.

In the meantime, it's still quite simple to customize this config:

```yaml
services:
  dms:
    configs:
      - source: oauth2
        target: /etc/dovecot/dovecot-oauth2.conf.ext

configs:
  oauth2:
    content: |
      client_id = dms-dovecot
      client_secret = hunter2
      introspection_mode = post
      introspection_url = https://auth.example.com/admin/oauth2/introspect
```

The docs page could be improved further with guidance for such, I would appreciate users of the feature sharing the different config differences they need.

**Comment by polarathene** [comment]
I also want to ask what about your setup requires introspection? Is `/userinfo` endpoint not sufficient? That requires no credentials as it's a public endpoint.

I know that you use and contribute to Roundcube and that it's often mentioned by users with DMS so I've been trying to get that setup along with an actual OAuth2 service that I can host locally.

Presently I've tried with:
- Ory Hydra (_I can acquire an access token via their consent ui demo, but I'll either need to modify that demo a bit or extend it further with Ory Kratos for actually user data to be present with requested `email` and `profile` scopes that presently add nothing_).
- Kanidm (_this one has been a bit frustrating to setup, their docs are a tad inaccurate and some security practices such as enforced TLS/HTTPS adds some friction, but I think I'm further along with this one_)
- Rauthy (_current attempt_)
- Authentik (_next attempt if rauthy doesn't help, avoided this initially because no embedded sqlite option, requires additional services added postgresql and something else IIRC_)

---

Ideally I want one of these with minimal config to bootstrap a demo with DMS + Roundcube:

```yaml
services:
  dms:
    image: mailserver/docker-mailserver:edge
    hostname: mail.example.test
    environment:
      # Enable the OAuth2 support in Dovecot and configure it for our mocked service (caddy):
      ENABLE_OAUTH2: 1
      OAUTH2_INTROSPECTION_URL: https://auth.example.test/oauth2/token/introspect/
    configs:
      - source: dms-accounts
        target: /tmp/docker-mailserver/postfix-accounts.cf

  # NOTE: Other containers can reach each other via the service name, eg http://roundcube
  roundcube:
    image: roundcube/roundcubemail:latest
    # hostname: rc.example.test
    ports:
      - 80:80
    environment:
      - ROUNDCUBEMAIL_DB_TYPE=sqlite
      - ROUNDCUBEMAIL_SKIN=elastic
      - ROUNDCUBEMAIL_DEFAULT_HOST=tls://mail.example.test
      - ROUNDCUBEMAIL_SMTP_SERVER=tls://mail.example.test
    configs:
      - source: rc-oauth2
        target: /var/roundcube/config/oauth2.inc.php

# The `configs` compose feature works with any release of Docker Compose since 2024.
# NOTE: `$` needed to be repeated to escape it, which opts out of the `compose.yaml` variable interpolation feature.
configs:
  # Both users can login via roundcube interface full email address below as the username and password `secret`
  dms-accounts:
    content: |
      john.doe@example.test|{SHA512-CRYPT}$$6$$sbgFRCmQ.KWS5ryb$$EsWrlYosiadgdUOxCBHY0DQ3qFbeudDhNMqHs6jZt.8gmxUwiLVy738knqkHD4zj4amkb296HFqQ3yDq4UXt8.
      jane.doe@example.test|{SHA512-CRYPT}$$6$$sbgFRCmQ.KWS5ryb$$EsWrlYosiadgdUOxCBHY0DQ3qFbeudDhNMqHs6jZt.8gmxUwiLVy738knqkHD4zj4amkb296HFqQ3yDq4UXt8.

  rc-oauth2:
    content: |
      <?php

      $$config['oauth_provider'] = 'generic';
      $$config['oauth_provider_name'] = 'Test OAuth2 Provider';
      $$config['oauth_client_id'] = 'roundcube';
      $$config['oauth_client_secret'] = '8NtsWYusyUPsbSQvXf95kGkSKBDaZs8SBTqkU4Dw2Gz9gx5E';

      // This appears to be used only by the browser URL? Hence without custom DNS I've used this to connect to the auth container directly:
      // If roundcube does try to use it internally, obviously it will fail to route to the auth container.
      $$config['oauth_auth_uri'] = 'https://localhost:8443/ui/oauth2/authorise';

      $$config['oauth_token_uri'] = 'https://auth.example.test/oauth2/token';
      $$config['oauth_identity_uri'] = 'https://auth.example.test/oauth2/token/introspect';

      // Optional: disable SSL certificate check on HTTP requests to OAuth server. For possible values, see:
      // http://docs.guzzlephp.org/en/stable/request-options.html#verify
      $$config['oauth_verify_peer'] = false;

      $$config['oauth_scope'] = 'email openid profile';
      $$config['oauth_identity_fields'] = ['email'];

      // Boolean: automatically redirect to OAuth login when opening Roundcube without a valid session
      $$config['oauth_login_redirect'] = false;

      ?>
```

With kanidm I have the roundcube container offering a login screen with the OAuth2 login button, that redirects to kanidm to login if needed, and then the roundcube redirect URL as the roundcube OAuth2 wiki guide mentions is used, however this just redirects me back to the kanidm logged in page instead of the roundcube interface I get if I use the username/pass credentials without OAuth2.

I am not sure if I've misconfigured something in roundcube or kanidm (_hence my attempt to verify by getting alternative setup_), or if it's another issue (roundcube container difference? caddy as a reverse proxy in front of roundcube?).

Any assistance there would be appreciated. I've not used roundcube before, but I know that it connects to DMS with the above just fine. Regular login would show emails I sent to the account mailbox. Just need to fix the oauth2 login process.

**Comment by Westie** [comment]
> Before you do, what exactly do you want to PR?

After some reflection and further debugging, I really don't actually quite know.

It ultimately turns out that the entire root of my issues (including the requirement to add `skip` to passdb and userdb entries) were to do with the fact that I had my ldap passdb set up to authenticate basically everything, because I had omitted the `mechanism` config entry, like so:

```
passdb {
    driver = ldap
    args = /etc/dovecot/dovecot-ldap.conf.ext
    skip = authenticated
}
```

It turns out that the real fix was to do something like this instead:

```
passdb {
    driver = ldap
    args = /etc/dovecot/dovecot-ldap.conf.ext
    mechanism = plain login none
}
```

Note the addition of `none`, which allows the LDAP passdb driver to be used to query non-authenticating requests such as LMTP etc.

Finally removing `skip` worked too, so I wonder if all that's needed is substituting `mechanism = plain login none` into the core `auth-ldap.conf.ext` file?

But I'm _very_ curious how you managed to get LMTP working without any non-authenticating passdb lookups on your local environment, for example, I don't have any local cache of users at all on my environment, it's all entirely seeded from ldap

This is all very much a trial and error exercise as I haven't properly worked on dovecot for well over 3 years at this point 😂 

And of course this applies to me when finishing off migrating to the latest version, gah!

![](https://ss.westie.sh/OFI2m)

**Comment by polarathene** [comment]
> After some reflection and further debugging, I really don't actually quite know.

Fair enough. I've found the roundcube and dovecot docs to be a bit vague in this area myself.

I think it's more of a documentation issue that we can better address on our end once enough users chime in with experiences so we know about common config needs and edge cases 👍 

---

> Note the addition of `none`, which allows the LDAP passdb driver to be used to query non-authenticating requests such as LMTP etc.

Do you have more info regarding this? Why would the passdb be used for such? What queries?

`mechanisms` was explicitly configured here in DMS for support of OAuth2 as an additional passdb: https://github.com/docker-mailserver/docker-mailserver/pull/3812

I would like to get around to refactoring a split between userdb and passdb, allowing for an easier way to opt-out of the passdb if you're only interested in oauth2 authentication (_this would however prevent mail clients like Thunderbird from working, unless one of the hard-coded providers they allow is used, but is fine if everyone uses webmail like roundcube_).

UserDB is required for any lookups related to users outside of authentication. Both Postfix and Dovecot will still be interested in that, such as for delivering mail to a mailbox.

In future we may eventually have the equivalent of OAuth2 auth for user provisioning, SCIM which would keep identity management external of DMS with your identity provider instead (_perhaps less of a concern with LDAP vs the FILE `ACCOUNT_PROVISIONER` though_).

---

> I'm very curious how you managed to get LMTP working without any non-authenticating passdb lookups on your local environment

- Dovecot OAuth2 logins receive a username and password like a typical login, but the process takes an Access Token (_potentially a JWT ID Token, rather than an opaque string_) as the password to validate against the auth service. If successful then Dovecot attempts to match a field in the JSON response against the login username and if that matches authentication is successful.
- **EDIT:** While I recall in the past statements about not needing userdb, if the account wasn't known in DMS it does seem to fail the IMAP login/authentication process (_as per `curl` example in our OAuth `:edge` docs for upcoming DMS v14.1_).

Regarding LMTP, I don't think passdb has any relevance there? userdb does as Postfix will want to know the user is valid/exists (has a mailbox), and Dovecot will want to know where that mailbox should store mail.

> And of course this applies to me when finishing off migrating to the latest version, gah!

Could you clarify what you need `/introspect` post for over `/userinfo`? Are you using an open-source auth service that requires it to work?

I am not sure if rauthy has a proper introspection endpoint, I might need to try with an alternative that does. I didn't seem to get a different result between DMS v13.3 or v14 despite that changelog entry for Dovecot.

```
dovecot: auth: Error: oauth2(jane.doe@example.test,172.16.42.5,<4OlUf9seAIKsECoF>): oauth2 failed: Introspection failed: Object doesn't begin with '{'
```

Yet when I query the same rauthy endpoint (`/tokenInfo`) via Insomnia with a JSON `{ "token": "id-token-goes-here" }` I would get a response:

```json
{
	"active": true,
	"scope": "openid email profile",
	"client_id": "roundcube",
	"username": "za9UxpH7XVxqrtpEbThoqvn2",
	"exp": 1722778258
}
```

The `username` field is for the `jane.doe@example.test` identity in rauthy, but rauthy has created it with a random value as an ID. If I instead query the `/userinfo` endpoint with the ID Token via `Authorization: Bearer id-token-goes-here` header, I get:

```json
{
	"id": "za9UxpH7XVxqrtpEbThoqvn2",
	"sub": "za9UxpH7XVxqrtpEbThoqvn2",
	"name": "Rauthy Admin",
	"roles": [
		"rauthy_admin",
		"admin"
	],
	"mfa_enabled": false,
	"email": "jane.doe@example.test",
	"email_verified": true,
	"preferred_username": "jane.doe@example.test",
	"given_name": "Rauthy",
	"family_name": "Admin",
	"locale": "en"
}
```

Even on DMS v14 I could use basic auth within the `introspection_uri` and the reverse proxy (Caddy) showed in logs that it was using Basic Auth. `client_secret` and `client_id` didn't seem to influence that with v13.3.

**EDIT:**
- `/tokenInfo` endpoint for Rauthy only seems to accept a request with `Content-Type` of `application/json` while Dovecot was sending with `application/x-www-form-urlencoded`, which is why a JSON response wasn't being returned causing Dovecot to error.
- Earlier observations mentioned in this comment also lacked insight into the request body, Caddy's logs don't include that so there's probably a difference going on there between DMS v13.3 and DMS v14 🤷‍♂️

**Comment by Westie** [comment]
Hello again @polarathene, it is me, your ever-absent OIDC advocate, once again taking many months to respond to an excellent post of yours. apologies for the delay, life unfortunately continues to get in the way.

Let's attempt to answer some questions you asked!

## Why introspect? Why not userinfo?
It turns out that this is something that is ultimately specific to a particular configuration. Various posts on reddit suggest that the [introspection URL is the correct method](https://www.reddit.com/r/selfhosted/comments/omwb2j/any_one_get_dovecot_keycloak_working_for_with/) to use if you're a Keycloak user like me.

However as far as I recall, it is now possible to create a userinfo mapping in Keycloak but that may require extra configuration which to me is extra effort for little, if any gain.

This is something that the [dovecot docs appears to suggest is the case](https://doc.dovecot.org/2.3/configuration_manual/authentication/oauth2/) as well without explaining why or pointing out differences, it just suggests that Google and WSO2 use different flows.

Therefore, I think for anyone attempting to integrate OIDC into Dovecot, it's a case of trial and error to figure out whether or not userinfo (preferred, in most cases imho) or introspect is the correct flow for authenticating requests.

## The issue with LMTP
When I was debugging this previously in the year, I had experienced issues with regards to local e-mail resolution, where no local users exist.

To describe what was happening:

1. `user1@lol.com` attempts to send an e-mail to `group1@lol.com`
2. `group1@lol.com` includes the following members: `user1@lol.com`; `user2@lol.com`
3. dovecot attempts to query for users `user1@lol.com` and `user2@lol.com`
4. since those two users do not exist locally, errors are thrown and e-mails aren't sent to group members

But for some reason, when I got around to debugging this further today, the error was no longer being thrown. I wonder if I had somehow just upset my local mailserver.

Sometimes debugging really is like chasing ghosts.

**Comment by hadestructhor** [comment]
Hello everyone, i'm joining this conversation as I've been trying to setup dms and roundcube with casdoor as an authentication provider using OIDC.

I managed after long days of debugging and inspecting to find that casdoor wasn't returning the correct response, and after a few issues raised in their side with the detail about the problem I was having authenticating users, I finally managed to get past the authentication point.

If I understand correctly, I have to create accounts manually for the OIDC connection to work as the authentication and user databases are seperate. 

Is there any way to provision users into the userdb ? I tried looking online but I can't find much about it. 

For anyone that's interested, I've linked my configuration for dms, roundcube and casdoor
[casdoor.zip](https://github.com/user-attachments/files/18025231/casdoor.zip)

**Comment by NhProGamer** [comment]
In the past, I had a lot or issues with casdoor, their implementation of the oidc instant good, use authentik, authelia or logto, these alternatives are just better

**Comment by hadestructhor** [comment]
> In the past, I had a lot or issues with casdoor, their implementation of the oidc insn't good, use authentik, authelia or logto, these alternatives are just better

Fair enough, I saw they had some specific sections not implementing the correct standards that caused dovecot to fail login the users, but for after opening issues on their repository, I now managed to make the authentication part work. 

I'll look into the other alternatives at another time as right now I'm more interested in provisioning user accounts in DMS after a successful authentication if the user did not already exist.

**Comment by Westie** [comment]
> Hello everyone, i'm joining this conversation as I've been trying to setup dms and roundcube with casdoor as an authentication provider using OIDC.

Don't forget that OIDC is for authenticating users, it shouldn't be used as the source of truth for whether or not a user itself exists.

You would need to use something like the [LDAP Server](https://casdoor.org/docs/ldap/ldapserver/) or create your own mechanism for syncing across users and groups.

**Comment by polarathene** [comment]
**NOTE:** I've got too much backlog presently to engage much on this topic atm. Hope this overview helps.

I had been delving into this topic earlier in the year and some config examples and caveats I wanted to PR into DMS, but other priorities have gotten in the way of that 😓 

While verbose, I have documented some relevant info across issues:
- https://github.com/docker-mailserver/docker-mailserver/issues/4081#issuecomment-2179754501
- https://github.com/docker-mailserver/docker-mailserver/issues/4119#issuecomment-2270226402
- https://github.com/docker-mailserver/docker-mailserver/issues/3830

---

> If I understand correctly, I have to create accounts manually for the OIDC connection to work as the authentication and user databases are seperate.
> 
> Is there any way to provision users into the userdb ? I tried looking online but I can't find much about it.

Correct.
- The OIDC/OAUTH2 support is purely authentication as an alternative to a traditional password.
- Account provisioning is a separate step you should take care of prior as it's required for mail to be delivered into a mailbox at `/var/mail`.

DMS `ACCOUNT_PROVISIONER` ENV presently supports values of `FILE` and `LDAP`.
- There are plans for SCIM eventually which would allow a compatible IdP to sync/provision accounts (_DMS would expose a SCIM endpoint which provides a standardized HTTP API for an IdP to manage the related account config for DMS_).
  - https://github.com/docker-mailserver/docker-mailserver/issues/4121
  - https://github.com/docker-mailserver/docker-mailserver/issues/4090
- Some IdP will offer LDAP, which you might be able to configure for DMS, but our default query filters for LDAP are configured with OpenLDAP + Postfix schema extension in mind.
- There is also some third-party community projects with web UI for managing accounts that is compatible with `ACCOUNT_PROVISIONER=FILE`. I'm not sure if they offer a separate API service that you could likewise leverage, but these projects tend to map that API to calling the equivalent method in our `setup` CLI command in the container.

---

For Roundcube, and also relevant to Dovecot IIRC, there are some configuration concerns I recall when I was looking into DMS + OAuth2 support. I thought I had documented that somewhere online (_at least the username case config, but I'm not sure where_ 😓 )

I'll share some config to reference if it helps, but I don't have time right now to verify it runs properly (_nor correct it if it does not_).

**Comment by polarathene** [comment]
This is a revised version of the earlier comment above [`compose.yaml` for DMS + Roundcube with OIDC](https://github.com/docker-mailserver/docker-mailserver/issues/2713#issuecomment-2267147187):

## DMS + Roundcube

**`mail.compose.yaml`:**

```yaml
# Make sure you run with `RC_OAUTH2__CLIENT_SECRET` ENV set (or assigned in a `.env` next to your `compose.yaml`)
# This ENV is referenced below in several places and will be interpolated into the config.

# The custom network is only relevant with the separate reverse proxy (Caddy) config,
# - It was separated along with the IdP for easier management/testing.
# - This network provides an alias `auth.example.test` to the reverse proxy to handle, any containers that would see that FQDN would then know to direct the request to the reverse proxy. This was for local/offline testing.
# - You'll also see `auth.example.localhost` referenced which relies on the Docker host to resolve any `.localhost` FQDN to localhost, which with the reverse proxy publishing ports 80 and 443 on a desktop system would allow you to use your host OS apps to connect.

networks:
  default:
    name: custom-network
    external: true

services:
  dms:
    image: mailserver/docker-mailserver:edge
    hostname: mail.example.test
    # Optional This is to support HTTPS for a private CA cert via `user-patches.sh`:
    volumes:
      - ./certs/:/tls/example/:ro
    environment:
      # Enable the OAuth2 support in Dovecot
      ENABLE_OAUTH2: 1
      #OAUTH2_INTROSPECTION_URL: https://auth.example.test/api/oidc/userinfo
    configs:
      # Q: UserDB must have accounts matching the `email` claim?
      - source: dms-accounts
        target: /tmp/docker-mailserver/postfix-accounts.cf
      # Optional - Custom config override when our ENV support is lacking (DMS v14 has limited ENV support):
      - source: oauth2-config
        target: /etc/dovecot/dovecot-oauth2.conf.ext
      # Optional:
      - source: dovecot-cf
        target: /tmp/docker-mailserver/dovecot.cf
      # Add custom CA to the trust store (local IdP container has HTTPS with self-signed cert, not LetsEncrypt):
      # - Dovecot lacks support to ignore cert verification when connecting over HTTPS.
      # - Some IdP enforce HTTPS (at least for `/introspect` endpoint, `/userinfo` doesn't require it).
      - source: user-patches
        target: /tmp/docker-mailserver/user-patches.sh

  roundcube:
    image: roundcube/roundcubemail:latest
    environment:
      - ROUNDCUBEMAIL_DB_TYPE=sqlite
      - ROUNDCUBEMAIL_SKIN=elastic
      - ROUNDCUBEMAIL_DEFAULT_HOST=tls://mail.example.test
      - ROUNDCUBEMAIL_SMTP_SERVER=tls://mail.example.test
    configs:
      - source: rc-oauth2
        target: /var/roundcube/config/oauth2.inc.php

# This example uses the Docker Compose `configs` feature instead of separate files so you can copy/paste this `compose.yaml`
# - The feature is compatible with any Compose release since 2024.
# - The character `$` below is changed to `$$` to opt-out of accidental variable interpolation (Compose feature).
configs:
  # An account provisioned for you already:
  # - Username: john.doe@example.test
  # - Password: secret
  dms-accounts:
    content: |
      john.doe@example.test|{SHA512-CRYPT}$$6$$sbgFRCmQ.KWS5ryb$$EsWrlYosiadgdUOxCBHY0DQ3qFbeudDhNMqHs6jZt.8gmxUwiLVy738knqkHD4zj4amkb296HFqQ3yDq4UXt8.
      jane.doe@example.test|{SHA512-CRYPT}$$6$$sbgFRCmQ.KWS5ryb$$EsWrlYosiadgdUOxCBHY0DQ3qFbeudDhNMqHs6jZt.8gmxUwiLVy738knqkHD4zj4amkb296HFqQ3yDq4UXt8.

  # Optional - Install the custom CA cert into trust store:
  user-patches:
    content: |
      cp /tls/example/ca-cert.pem /usr/local/share/ca-certificates/smallstep-ca.crt
      update-ca-certificates

  # Optional
  # - `auth_username_format` affects format of the login username received from mail client
  # - Debug settings may help when troubleshooting your integration.
  dovecot-cf:
    content: |
      auth_username_format = %u
      auth_debug=yes
      auth_debug_passwords=yes

  # Refer to these notes until DMS docs are updated:
  # https://github.com/docker-mailserver/docker-mailserver/issues/4081#issuecomment-2179754501
  oauth2-config:
    content: |
      # Prefer `/userinfo` which tends to be a public endpoint (not as privileged as `/introspect`)
      introspection_url = https://auth.example.test/api/oidc/userinfo

      # Actual introspection endpoint tends to be POST request with credentials.
      #introspection_url = https://auth.example.test/api/oidc/introspect
      #introspection_mode = post

      # NOTE: DMS v14 basic auth support regressed compared to DMS v13.
      # This is due to the Dovecot version regressing (no community build for Debian 12 available):
      # https://github.com/docker-mailserver/docker-mailserver/issues/2713#issuecomment-2268187285
      # https://github.com/docker-mailserver/docker-mailserver/issues/4081#issuecomment-2267152565

      #client_id = roundcube
      #client_secret = ${RC_OAUTH2__CLIENT_SECRET}

      # Alternative syntax (DMS v14 compatible) for supplying introspection with credentials via basic auth:
      #introspection_url = https://roundcube:${RC_OAUTH2__CLIENT_SECRET}@auth.example.test/api/oidc/introspect

      # Adjust which attribute from the JSON response is chosen for getting the username from,
      # this will be used to compare with the login username Dovecot received
      #username_attribute = preferred_username
      #username_attribute = username
      
      # Dovecot formatting tokens to modify/normalize the username returned from login
      # to match the IdP username provided via JSON response:
      #username_format = %Ln
      #username_format = %n

  # Basic OAuth2 / OIDC config example with several IdP:
  rc-oauth2:
    content: |
      <?php

      $$config['oauth_provider'] = 'generic';
      $$config['oauth_provider_name'] = 'Test OIDC Provider';
      $$config['oauth_client_id'] = 'roundcube';
      $$config['oauth_client_secret'] = '${RC_OAUTH2__CLIENT_SECRET}';

      // oauth_auth_uri is a redirect URI configured in both roundcube here and your IdP
      // It uses localhost instead of test for the TLD as this example is intended to be run
      // from your desktop system. It is compatible with Linux and Windows (WSL2).

      // kanidm
      //$$config['oauth_auth_uri'] = 'https://auth.example.localhost/ui/oauth2';
      //$$config['oauth_token_uri'] = 'https://auth.example.test/oauth2/token';
      //$$config['oauth_identity_uri'] = 'https://auth.example.test/oauth2/openid/roundcube/userinfo';

      // rauthy
      //$$config['oauth_auth_uri'] = 'https://auth.example.localhost/auth/v1/oidc/authorize';
      //$$config['oauth_token_uri'] = 'https://auth.example.test/auth/v1/oidc/token';
      ////$$config['oauth_identity_uri'] = 'https://auth.example.test/auth/v1/oidc/introspect';
      //$$config['oauth_identity_uri'] = 'https://auth.example.test/auth/v1/oidc/userinfo';

      // authelia
      $$config['oauth_auth_uri'] = 'https://auth.example.localhost/api/oidc/authorization';
      $$config['oauth_token_uri'] = 'https://auth.example.test/api/oidc/token';
      //$$config['oauth_identity_uri'] = 'https://auth.example.test/api/oidc/introspect';
      $$config['oauth_identity_uri'] = 'https://auth.example.test/api/oidc/userinfo';

      // Optional: disable SSL certificate check on HTTP requests to OAuth server. For possible values, see:
      // http://docs.guzzlephp.org/en/stable/request-options.html#verify
      $$config['oauth_verify_peer'] = false;

      $$config['use_https'] = true;
      // Force normalization of login to lower-case:
      $$config['login_lc'] = 1;

      $$config['oauth_scope'] = 'email openid profile';
      $$config['oauth_identity_fields'] = ['email'];

      // Boolean: automatically redirect to OAuth login when opening Roundcube without a valid session
      $$config['oauth_login_redirect'] = false;

      ?>
```

The `compose.yaml` example is taken from what I found on disk from August 2024 with some additional commentary I've added for context based on what I recall for settings.

The username formatting ones specifically are a bit muddy in my memory, but you have to account for SASL login username, Dovecot UserDB (or LDAP username), the username that the IdP provides from it's own `/userinfo` / `/introspect` endpoint and also the one with Roundcube. This also may have had some relevance with DMS and `/var/mail` / `postfix-accounts.cf` IIRC where you may need to ensure that there is consistent in formatting, at the very least with casing, so normalizing to lower-case can be important 😅 

I know with RAuthy it's user management wasn't as flexible which is where I ran into a long debug session to find where each service was also affected by login usernames that weren't lower-case.

---

## Related (reverse proxy + IdP)

I have not included the IdP setup here, but if you'd like me to share that too let me know.

**`compose.yaml`:**

```yaml
# This `compose.yaml` creates the custom network and imports services defined from separate compose files:
# - Reverse Proxy
# - Identity Provider (IdP)

include:
  - rp.compose.yaml
  - authelia.compose.yaml

networks:
  default:
    name: custom-network
    ipam:
      driver: default
      config:
        - subnet: "172.16.42.0/24"
```

**`rp.compose.yaml`:**

```yaml
services:
  reverse-proxy:
    image: caddy:2.8
    volumes:
      - ./certs/:/tls/example/:ro
      - ./Caddyfile:/etc/caddy/Caddyfile:ro
    ports:
      - "80:80"
      - "443:443"
    networks:
      default:
        # This alias is to enable other services on this network to resolve
        # the IP of this container. More reliable than `links` / `extra_hosts`?
        # At least with links, you may need to define it per service, vs once?
        aliases:
          - auth.example.test
```

**`Caddyfile`:**

```Caddyfile
{
  # Forcing self-signed instead of requesting public certs:
  local_certs
  # For testing, disables HTTP => HTTPS, allowing you to troubleshoot
  # against HTTP only endpoints when supported by the IdP:
  auto_https disable_redirects

  # Only for debugging:
  servers {
    log_credentials
  }

  # Optional log format:
  log {
    format console {
      time_format "15:04:05"
    }
  }
}

# HTTP only roundcube for testing:
http://rc.example.localhost {
  reverse_proxy roundcube:80
}

# HTTPS only roundcube, tls directive refers to a certificate that was
# provisioned for wildcard domains `*.example.localhost` + `*.example.test`
https://rc.example.localhost {
  tls /tls/example/cert.pem /tls/example/key.pem
  reverse_proxy roundcube:80
}

# For your IdP adjust the container service name and port accordingly
# There are two FQDNs for routing to the container to support a local/offline testing environment:
# - `*.example.localhost` is for requests from your Docker host (mail client, web browser),
# - `*.example.test` is for within the custom Docker network
http://auth.example.localhost, http://auth.example.test {
  reverse_proxy authelia:9091
}

# Kanidm example (NOTE: I would discourage choosing this IdP)
# The kanidm service certificate in this case is the exact same one
# referenced by the Caddy tls directive.
https://auth.example.localhost, https://auth.example.test {
  tls /tls/example/cert.pem /tls/example/key.pem

  # Kanidm does not allow listening on HTTP, it enforces HTTPS even with a proxy,
  # Ensure it receives the Host header it expects for it's configuration and
  # establish HTTPS again with SNI set to it's own provisioned certificate.
  reverse_proxy https://kanidm-server:8443 {
    header_up Host auth.example.localhost
    transport http {
      tls_server_name auth.example.localhost
      # If Caddy would not have the CA for kanidm in trust store, skip verify:
      tls_insecure_skip_verify
    }
  }
}
```

**Comment by hadestructhor** [comment]
> > Hello everyone, i'm joining this conversation as I've been trying to setup dms and roundcube with casdoor as an authentication provider using OIDC.
> 
> @hadestructhor Don't forget that OIDC is for authenticating users, it shouldn't be used as the source of truth for whether or not a user itself exists.
> 
> You would need to use something like the [LDAP Server](https://casdoor.org/docs/ldap/ldapserver/) or create your own mechanism for syncing across users and groups.

It's me again @Westie. I've read through the ldap configuration page of docker mailserver and configured it to match the ldap exposed through casdoor. 

I managed to search the ldap via an old ldap application I found so I know the users do exist.

Now I'm struggling getting it to work with docker mailserver, specifically dovecot which gives me the following error:

```stacktrace
2024-12-09T16:30:35.601500+01:00 mail dovecot: auth: Error: ldap(test@domain.tld,172.21.0.3,<w4KyC9goRKisFQAD>): ldap_search($("(&(objectClass=posixAccount)(cn=test))")) parsing failed: Bad search filter
``` 

This is the current search filter I have in place :
```env
LDAP_START_TLS=                                                                                                                                                                                       
LDAP_SERVER_HOST=ldap://casdoor.domain.tld                                                                                                                        
LDAP_SEARCH_BASE=ou=Organization,dc=domain,dc=tld                                                                                                                    
LDAP_BIND_DN=cn=admin,dc=example,dc=com                                                                                                                                                       
LDAP_BIND_PW="secret"                                                                                                                        
LDAP_QUERY_FILTER_USER=$("(&(objectClass=posixAccount)(cn=%n))")                                                                                                                                      
LDAP_QUERY_FILTER_GROUP=                                                                                                                                                                             
LDAP_QUERY_FILTER_ALIAS=                                                                                                                                                                             
LDAP_QUERY_FILTER_DOMAIN=$("(mail=*@%s)")                                                                                                                           
LDAP_QUERY_FILTER_SENDERS=$("(|(mail=%s)(mail=admin@*))")
                                                                                                                                                   
DOVECOT_TLS=                                                                                                                                                                   
DOVECOT_USER_FILTER=$("(&(objectClass=posixAccount)(cn=%n))")                                                                                                                  
DOVECOT_PASS_FILTER=$("(&(objectClass=posixAccount)(cn=%n))")                                                                                                       
DOVECOT_USER_ATTRS=home=%{ldap:homeDirectory},uid=%{ldap:uidNumber},gid=%{ldap:gidNumber},mail=maildir:/var/spool/vmail/%{ldap:cn}/Maildir                     
DOVECOT_MAILBOX_FORMAT=maildir                                                                                                                                      
DOVECOT_AUTH_BIND=yes

ENABLE_SASLAUTHD=1                                                                                                                                                                            
SASLAUTHD_MECHANISMS=ldap                                                                                                                                           
```

I tried replacing the filter with `(uid=%u)` and `(uid=%n)` but neither work.

I’m certain the ldap connection works as I was getting rate limited attempts with clear errors from my casdoor server, and now I don’t. 

Am I maybe configuring the filters wrong ?

**Comment by netliam** [comment]
> Here's my take on the subject. I'm not really a mail expert, but given the short amount of time I had to find a working solution, I thought it would be interesting to share my thoughts and solutions, in order to help others (and maybe contribute to the `OIDC` provider development).
> ## Enable OIDC (OAuth2) authentication in Docker Mailserver
> 
> Following your comments here and some guides such as https://doc.dovecot.org/configuration_manual/authentication/oauth2/ you can easily configure Dovecot to accept authentication from your OIDC/OAuth2 provider (in my case, Keycloak)
> 
> However, at this point **account login will only work if you previously created this account** (I'm using `ACCOUNT_PROVISIONER=FILE` btw). So we need something that will be able to create accounts (= fill `postfix-accounts.cf` and wait for the `changedetector` to trigger `accounts.sh` & all)
> ## Effectively create accounts
> ### Alternative 1 - Using a static userdb
> 
> Using the `userdb` config for `passwd-file` and the configuration provided at https://documentation.open-xchange.com/7.10.4/middleware/mail/dovecot/oauth_2.0_with_postfix_and_dovecot.html we added a `userdb` entry to the `oauth2` config file. Doing so, we were able to connect successfully (using Roundcube for example), and we could send emails, but that new account couldn't be found by others (some files were missing because not created, obviously).
> 
> `/etc/dovecot/conf.d/auth-oauth2.inc`
> 
> ```diff
> +userdb {
> +  driver = static
> +  args = username_format=%u uid=docker gid=docker home=/var/mail/%d/%u
> +  default_fields = uid=docker gid=docker home=/var/mail/%d/%u
> +}
> ```
> 
> **This first alternative is not fully satisfying because it does not setup everything correctly.**
> ### Alternative 2 - Using post-login scripting
> 
> I believe we could use dovecot's [post-login scripting](https://doc.dovecot.org/admin_manual/post_login_scripting/) to trigger a modification inside the `postfix-accounts.cf` with a simple shell script and wait for the folder/files to be created.
> 
> **This second alternative feels satisfying.** An implementation to validate it would be interesting. However, that raise a new issue : **mail accounts/boxes will not exist prior to the first login of each user** 🤯 So you will not be able to receive any email, even if your account exists in your OIDC Provider, because on the mailserver side, nothing exists.
> ### Alternative 3 - External service
> 
> In order to adress the two issues (accounts need to be created to be able to login, accounts need to be created to receive emails on existing accounts before login), another alternative is to use an external service/trigger/hook, that would (regularly or on each new user creation) get existing users from the provider and create them in mailserver using `setup email add`.
> 
> **That's not perfect because it's an hybrid mode (between `FILE` and the goal for `OIDC` I believe), and it requires an external custom component, but that can solve (specific, I know) issues.**
> ## The end
> 
> Do not hesitate to give me your thoughts on these elements, and correct me if I'm wrong 😉

I came up with another solution that might be easier to implement, the downside is that it would need to be tweaked to work with the different SSO solutions. In my setup I'm using authentik so the HTTP request is tailored to authentik's API.

```Lua
local json = require "json"

local apikey = "apikeyhere"
local http_url = "https://authentik.company/api/v3/core/users/"
local http_method = "GET"

function script_init()
    return 0
end

function script_deinit()
end

function auth_userdb_lookup(req)
    local request = {email=req.user}

    local auth_request = http_client:request (
        url = http_url;
        method = http_method;
    }
    auth_request:add_header("Authorization", "Bearer " .. apikey)
    auth_request:set_payload(request)
    local auth_response = auth_request:submit()
    local auth_response_payload = auth_response:payload()

    local person = auth_response_payload.results[0]

    if person.email == req.user then
        return dovecot.auth.USERDB_RESULT_OK, "uid=5000 gid=5000 home=/var/mail/" .. person.username .. "mail=maildir:~/Maildir"
    else
        return dovecot.auth.USERDB_RESULT_USER_UNKNOWN, ""
    end
end
```

**Comment by thechubbypanda** [comment]
There was talk about potentially integrating with the Lua scripts for OIDC in the original PR (I think) #3480.

It's a great starting point but we'd have to implement a script for every OIDC provider. More providers are starting to implement SCIM (suggested implementation in #4090) e.g. [Rauthy](https://github.com/sebadob/rauthy/issues/666)
Authentik already has support.

**Comment by boecks** [comment]
I just commented in #4090 though, I am not sure, which of the many places is the best to discuss on this.

As mention in my other post, I am afraid, one cannot auto-provision new users using setup-cli at all, since one must provide a password. As far as my understangin goes, this includes all kinds of sources, like hook-scripts, SCIM, etc. (except when you can manage to intercept the plaintext password at the time it is set or updated)

Given that, the post-login method looks most promising. 

Another option is to do a one-time login when the user is created in IdP and afterwards do a password reset.

Please correct me if I am wrong  - I did not yet fully grasp, how the dovecot user data (i.e. mail and password) interferes with the postfix user data. If a dovecot login "overrules" given postfix credentials, one could create the postfix account with a dummy password or the like.

**Comment by polarathene** [comment]
I [responded](https://github.com/docker-mailserver/docker-mailserver/issues/4090#issuecomment-2869288718) at #4090 for context.

If you only need `setup email add` support to be changed a little, we could do that. If you're delegating auth to an IDP instead, then you can probably just generate a random password to prevent it from being used (128-bit of entropy will suffice). Or we could instead allow disabling password based logins, making that detail irrelevant 🤔 

I'm also open towards a breaking change in future that changes out `postfix-accounts.cf` format to a more standard config format like TOML/JSON, but that might be some time away. It would make interacting with the file easier.

**Comment by sebadob** [comment]
> More providers are starting to implement SCIM (suggested implementation in https://github.com/docker-mailserver/docker-mailserver/issues/4090) e.g. https://github.com/sebadob/rauthy/issues/666

Just fyi, Rauthy added SCIM support in `v0.29` and it's available now.

**Comment by polarathene** [comment]
Thanks @sebadob ! I'm very much interested in giving rauthy a revisit when I have time to come back to this feature, I recall it being one of the nicer solutions compared to alternatives 😎 Great to hear you've added SCIM support too 😁
